lib/hanami/providers/db.rb

```diff
@@ -5,6 +5,12 @@ module Providers
     # @api private
     # @since 2.2.0
     class DB < Dry::System::Provider::Source
+      class << self
+        def connection_cache
+          @connection_cache ||= {}
+        end
```

Comment:
Have you considered using `Dry::Core::Cache` for this? I think that would alleviate both the thread-safety concern (it's based on `Concurrent::Map`), and also the extension problem.

```ruby 
class DB < Dry::System::Provider::Source
  extend Dry::Core::Cache
```

```ruby
gateway = fetch_or_store database_url, config.extensions do
  ROM::Gateway.setup(config.adapter, database_url, extensions: config.extensions)
end
```

This will give you a unique Gateway for every combination of url + extensions in a thread-safe cache.

I chose to cache the entire Gateway in this example because its only purpose is to initialize the adapter and apply extensions, and that would greatly simplify the cache logic.

Reply:
@alassek Great point, and yep, using something like that makes a lot of sense to me. 

We have a similar pattern of cache usage inside hanami/view that we could follow for this too.

(The reason I didn't use it yet is because I was hacking this together at 11pm and just wanted to see whether things even worked in the first place 😄)